Source/C++/Core/Ap4CommonEncryption.cpp

```diff
@@ -2359,6 +2359,20 @@ AP4_CencDecryptingProcessor::CreateTrackHandler(AP4_TrakAtom* trak)
                 sample_descs.Append(protected_desc);
                 sample_entries.Append(sample_entry);
             }
+
+            AP4_ContainerAtom* schi = protected_desc->GetSchemeInfo()->GetSchiAtom();
+            if (schi != NULL) {
+                AP4_CencTrackEncryption* tenc;
+                if (protected_desc->GetSchemeType() == AP4_PROTECTION_SCHEME_TYPE_PIFF) {
+                    tenc = AP4_DYNAMIC_CAST(AP4_CencTrackEncryption, schi->GetChild(AP4_ATOM_TYPE_UUID));
+                } else {
+                    tenc = AP4_DYNAMIC_CAST(AP4_CencTrackEncryption, schi->GetChild(AP4_ATOM_TYPE_TENC));
+                }
+                if (tenc != NULL) {
+                    const AP4_DataBuffer* key = m_KeyMap->GetKeyByKid(tenc->GetDefaultKid());
+                    if (key != NULL) m_KeyMap->SetKey(trak->GetId(), key->GetData(), 16);
```

Comment:
Writing to the key map from AP4_CencDecryptingProcessor breaks the constness of that map. It doesn't seem right. Can you explain why you need to write a key here?

Reply:
mp4decrypt couldn't properly handle selecting the track based on key ID, it only seemed to work by track ID.